packages/xstate-fsm/src/index.ts

```diff
@@ -170,22 +170,31 @@ export function createMachine<
       }
 
       if (stateConfig.on) {
-        const transitions = toArray(stateConfig.on[eventObject.type]);
+        const transitions: Array<
+          StateMachine.Transition<TContext, TEvent>
+        > = toArray(stateConfig.on[eventObject.type]);
 
         for (const transition of transitions) {
           if (transition === undefined) {
             return createUnchangedState(value, context);
           }
 
-          const { target = value, actions = [], cond = () => true } =
+          const { target, actions = [], cond = () => true } =
             typeof transition === 'string'
               ? { target: transition }
               : transition;
 
+          const isInternal = target === undefined;
+
           if (cond(context, eventObject)) {
-            const nextStateConfig = fsmConfig.states[target];
+            const nextStateConfig = fsmConfig.states[target ?? value];
```

Comment:
this is somewhat a dead code - the resulting variable won't ever be used if we have a target (I suspect that you have used `??` fallback only to satisfy types as this is now declared before it's actually used), so maybe to avoid this you could do this:
```ts
if (cond(context, eventObject)) {
  const nextStateConfig = fsmConfig.states[target ?? value];
  const allActions =
    target === undefined
      ? actions
      : ([] as any[])
          // Only add exit and entry actions
          .concat(stateConfig.exit, actions, fsmConfig.states[target].entry)
          .filter((a) => a)
          .map<StateMachine.ActionObject<TContext, TEvent>>((action) =>
            toActionObject(action, (machine as any)._options.actions)
          );
}
```

which would also differentiate `target === undefined` branching a little bit more explicitly